rclpy/rclpy/impl/logger_rcutils.py

```diff
@@ -0,0 +1,275 @@
+    }
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(cls, cls).initialize_context(context, **kwargs)
+        context['first_has_been_skipped'] = False
+
+    @staticmethod
+    def log_condition(context):
+        logging_condition = True
+        if not context['first_has_been_skipped']:
+            logging_condition = False
+            context['first_has_been_skipped'] = True
+        return logging_condition
+
+
+# The ordering of this dictionary matches the order in which filters will be processed.
+# Note(dhood): ordering in the constructor is only maintained as of Python 3.6,
+# so we assign the entries individually.
+supported_filters = OrderedDict()
+supported_filters['throttle'] = Throttle
+supported_filters['skip_first'] = SkipFirst
+supported_filters['once'] = Once
+
+
+def get_filters_from_kwargs(**kwargs):
+    detected_filters = []
+    all_supported_params = []
+    for supported_filter, filter_class in supported_filters.items():
+        filter_params = filter_class.params.keys()
+        all_supported_params.extend(filter_params)
+        if any(kwargs.get(param_name) for param_name in filter_params):
+            detected_filters.append(supported_filter)
+    # Check that all required parameters (with no default value) have been specified
+    for detected_filter in detected_filters:
+        for param_name, default_value in supported_filters[detected_filter].params.items():
+            if param_name not in kwargs:
+                if default_value is not None:
+                    kwargs[param_name] = default_value
+                else:
+                    raise RuntimeError(
+                        'required parameter "{0}" not specified '
+                        'but is required for the the logging filter "{1}"'.format(
+                            param_name, detected_filter))
+    for kwarg in kwargs:
+        if kwarg not in all_supported_params:
+            raise RuntimeError(
+                'parameter "{0}" is not one of the recognized logging options "{1}"'
+                .format(kwarg, all_supported_params)
+            )
+    return detected_filters
+
+
+class RcutilsLogger:
+
+    def __init__(self, name=''):
+        self.name = name
+        self.contexts = {}
+
+    def get_severity_threshold(self):
+        return _rclpy_logging.rclpy_logging_get_severity_threshold()
+
+    def set_severity_threshold(self, severity):
+        return _rclpy_logging.rclpy_logging_set_severity_threshold(severity)
+
+    def log(self, message, severity, **kwargs):
+        name = kwargs.pop('name', self.name)
+        return_log_condition = kwargs.pop('return_log_condition', False)
+
+        # Infer the requested log filters from the keyword arguments
+        detected_filters = get_filters_from_kwargs(**kwargs)
+
+        # Get/prepare the context corresponding to the caller.
+        caller_id = CallerId()
+        caller_id_str = pickle.dumps(caller_id)
```

Comment:
otherwise the object is different each time so using it as a key doesn't accomplish the history that's desired. @dirk-thomas suggested overwriting the `__hash__` method: I'll give that a shot